benchmarks/benchmarks.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2021 The Regents of the University of Michigan
+
+import random
+import string
+from itertools import islice
+from multiprocessing import Pool
+from tempfile import TemporaryDirectory
+
+from tqdm import tqdm
+
+import signac
+
+
+def _random_str(size):
+    return "".join(random.choice(string.ascii_lowercase) for _ in range(size))
+
+
+def _make_json_data(i, num_keys=1, data_size=0):
+    assert num_keys >= 1
+    assert data_size >= 0
+
+    data = {f"b_{j}": _random_str(data_size) for j in range(num_keys - 1)}
+    data["a"] = f"{i}{_random_str(max(0, data_size - len(str(i))))}"
+    return data
+
+
+def _make_job(project, num_keys, num_doc_keys, data_size, data_std, i):
+    size = max(0, int(random.gauss(data_size, data_std)))
+    job = project.open_job(_make_json_data(i, num_keys, size))
+    if num_doc_keys > 0:
+        size = max(0, int(random.gauss(data_size, data_std)))
+        job.document.update(_make_json_data(i, num_doc_keys, size))
+    else:
+        job.init()
+
+
+def generate_random_data(
+    project,
+    N,
+    num_keys=1,
+    num_doc_keys=0,
+    data_size_mean=0,
+    data_size_std=0,
+    parallel=True,
+):
+    assert len(project) == 0
+
+    if parallel:
+        with Pool() as pool:
+            p = [
+                (project, num_keys, num_doc_keys, data_size_mean, data_size_std, i)
+                for i in range(N)
+            ]
+            list(pool.starmap(_make_job, tqdm(p, desc="init random project data")))
+    else:
+        from functools import partial
+
+        make = partial(
+            _make_job, project, num_keys, num_doc_keys, data_size_mean, data_size_std
+        )
+        list(map(make, tqdm(range(N), desc="init random project data")))
+
+
+def setup_random_project(
+    N, num_keys=1, num_doc_keys=0, data_size_mean=0, data_size_std=0, seed=0, root=None
+):
+    random.seed(seed)
+    if not isinstance(N, int):
+        raise TypeError("N must be an integer!")
+
+    temp_dir = TemporaryDirectory()
+    project = signac.init_project(f"benchmark-N={N}", root=temp_dir.name)
```

Comment:
Any reason to not use (a potentially extended/adapted variant of) https://github.com/glotzerlab/signac/blob/e60bfbc579371698c2690605d10f88b3a12bc27c/signac/contrib/project.py#L2412 and https://github.com/glotzerlab/signac/blob/8a804d0e9b513e576ae622b9044914ff16d0ac9a/signac/testing.py#L8 here?

Reply:
The `TemporaryProject` yields a project when used as a context manager. The benchmark script requires something we can store (like `self.project` and `self.temp_dir`) in `setup` and then destroy during the `teardown` method. I'm not aware of a clean way to use that context manager here.

Reply:
Well, even if you don't use `TemporaryProject`, you could still use the `testing.init_jobs()` function?

Reply:
Hmm, I had forgotten about that function. It looks like it is currently designed for tests of projects with complex schemas, not tests of performance with varying data sizes. We would probably have to change that function to support the arguments like `num_keys`, `num_doc_keys`, `data_size`, `data_std`, etc., and that feels out of scope for this PR. I'll finalize and merge as-is, since you indicated this is a non-blocking issue.